File: C++ Tutorials/Classes/Inheritance/Single_Inheritance.cpp
#include <iostream>
using namespace std;

/*---------------------------------------------------------
In this section we will learn about the type of inheritance and will
see that how to work on them one by one.

1. Single inheritance
2. Multilevel inheritance
3. Multiple inheritance
4. Hierarchical inheritance
5. Hybrid inheritance

`1. Single Inheritance:` In Single Inheritance, a class is only
inherit from only one class. or in other word only one subclass is
inherited by one base class only.
---------------------------------------------------------*/

class Person
{

private:
    string name;
    int age;
    float height;

public:
    Person(string name, int age, float height)
    {
        // constructor of the Person class.
        this->name = name;
        this->age = age;
        this->height = height;
    }
};
/*---------------------------------------------------------*/

// Now we will inherit our Person into the Student class.

class Student : private Person
{
private:
    int id;
    string std;
    string subject;

public:
    Student(string name, int age,
            float height, int id, string std,
            string subject) : Person(name, age, height) // initiating the Person constructor.
    {
        this->id = id;
        this->std = std;
        this->subject = subject;
    }
};

int main()
{
    system("cls");

    // Note here Creating object of sub class will invoke
    // the constructor ob base classes.

    Student st("manish", 24, 5.7, 12345, "PG", "AI");

    return 0;
}

File: Java Classes/OOPS/src/Single_Inheritance.java
// single level inheritance
class Person {

    private String name;
    private int age;
    private double height;

    Person(String name, int age, double height){
        this.name = name;
        this.age = age;
        this.height = height;
        System.out.println("Person Constructor is called....");
    }

}

class Student extends Person{
    private int rollNo;
    private String subject;
    private String section;

    Student(String name, int age, double height, int rollNo, String subject,
            String section){

        super(name,age,height); // running Person constructor
//        super.name;

        this.rollNo  = rollNo;
        this.section = section;
        this.subject = subject;

        System.out.println("Student constructor is called....");
    }

    final void show(){
        System.out.println("person data");
    }

    final void show(String x){
        System.out.println(x + "! \t" + "Perosn Data");
    }
}
